odl/trafos/wavelet.py

```diff
@@ -834,14 +957,511 @@ def adjoint(self):
         if self.is_orthogonal:
             return self.inverse
         else:
-            # TODO: put adjoint here
-            return super().adjoint
+            if not isinstance(self.wbasis, pywt.Wavelet):
+                return InverseAdjBiorthWaveletTransform(dom=self.range,
+                                                        nscales=self.nscales,
+                                                        wbasis=self.wbasis)
+            else:
+                return super().adjoint
 
     @property
     def inverse(self):
         """The inverse wavelet transform."""
-        return WaveletTransform(dom=self.range, nscales=self.nscales,
-                                wbasis=self.wbasis, mode=self.mode)
+        if not isinstance(self.wbasis, pywt.Wavelet):
+            return BiorthWaveletTransform(dom=self.range, ran=self.domain,
+                                          nscales=self.nscales,
+                                          wbasis=self.wbasis)
+        elif isinstance(self.wbasis, pywt.Wavelet):
+            return WaveletTransform(dom=self.range, nscales=self.nscales,
+                                    wbasis=self.wbasis, mode=self.mode)
+
+
+class BiorthWaveletTransform(Operator):
+    """Discrete biorthogonal wavelet trafo between discrete L2 spaces."""
+
+    def __init__(self, dom, nscales, wbasis):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        dom : `DiscreteLp`
+            Domain of the biorthogonal wavelet transform (the "image domain").
+            The exponent :math:`p` of the discrete :math:`L^p`
+            space must be equal to 2.0.
+        nscales : `int`
+            Number of scaling levels in the wavelet transform.
+        wbasis : `str`
+            (``'JOSbiorthN'``) J.O. Stomberg's biorthogonal wavelets
+            with ``N`` referring to the filter length.
+            Possible values are:
+
+            1, 3, 5, 7, 9
+
+        Signal extention mode: reflexion
+        """
+
+        self.nscales = int(nscales)
+        self.wbasis = wbasis
+
+        if not isinstance(dom, DiscreteLp):
+            raise TypeError('domain {!r} is not a `DiscreteLp` instance.'
+                            ''.format(dom))
+
+        if dom.exponent != 2.0:
+            raise ValueError('domain Lp exponent is {} instead of 2.0.'
+                             ''.format(dom.exponent))
+
+        ndim = dom.ndim
+        if ndim == 1:
+            max_level = np.log2(dom.grid.shape[0])
+        elif ndim == 2:
+            max_level = np.log2(max(dom.grid.shape[0], dom.grid.shape[1]))
+        elif ndim == 3:
+            max_level = np.log2(max(dom.grid.shape[0], dom.grid.shape[1],
+                                dom.grid.shape[2]))
+        else:
+            raise NotImplementedError('ndim {} not 1, 2 or 3'
```

Comment:
Since the user never specified something like `ndim`, this error message will look cryptic. Better say something like 
`'dimension of input is {}, supported are 1, 2 and 3.'.format(dom.ndim)`. `len(dom.ndim)` will raise an error BTW.